Upgrade types for uuid@8.3.0

ctavan opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/uuidjs/uuid#api-summary
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Closes #46477.

RyanCavanaugh commented on types/uuid/interfaces.d.ts:
```diff
@@ -48,3 +48,17 @@ export interface v5Static {
+
+export type NIL = string;
+
+export enum Version {
```
This object doesn't exist at runtime; it should not be an enum. It'd actually be nice to remove this file entirely and fold it into the parent file, since there's no `uuid/interfaces` module

ctavan replied:
OK, you are right about both, the enum and inlining the interface definitions (I inherited that structure and – being rather new to TypeScript typings – never questioned it).

If I inline `uuid/interfaces` how do I go about only `export`ing the named exports that are actually exported by the `uuid` module? I'm getting

```
All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting.
```

errors if I don't export all types. I've seen https://stackoverflow.com/a/51970377/1053532 but what is preferred in DefinitelyTyped? Adding `export {}` or declaring a `namespace`?

Pointing me to a module where this is done correctly would be enough, I can then make the change.

ctavan replied:
@RyanCavanaugh friendly ping

ctavan replied:
FWIW I went for the `export {}` route.

ctavan commented:
I have inlined the type definitions into `index.d.ts` and removed `interfaces.d.ts` as suggested in the review. I've also removed the `Version` enum.

It would be great to prefer this PR over #46477 since the other PR contains some subtle differences with respect to return types that are not in line with how the authors of `uuid` define the API.

ctavan commented:
Just resolved merge conflicts that were introduced due to #46557

ctavan commented:
@rbuckton could you have a look at this one?

rbuckton commented on types/uuid/index.d.ts:
```diff
@@ -8,10 +8,73 @@
+type OutputBuffer = ArrayLike<number>;
+type InputBuffer = ArrayLike<number>;
 
+interface RandomOptions {
```
You might want to export these interfaces so that they are reachable, i.e.:

```ts
import { v1, V1Options } from "uuid";

const options: V1Options = {
  clockseq: 0,
  msecs: Date.now()
};

v1(options);
```

Otherwise anyone using the types would have to repeat the definitions themselves since they're not exported.

ctavan replied:
In this PR I was focusing on adding the new APIs that the upstream `uuid` module added in the latest version. The options objects have so far not been exposed, so while I agree that this is probably nice to have, it's at least not a regression with respect to the current state of the types.

Could we maybe first get this merged since people are waiting for it https://github.com/uuidjs/uuid/issues/502 and keep submitting incomplete PRs for these missing APIs https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46477 & https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46557?

I'm of course happy to submit the proposed improvements in another PR, but I'd really prefer to get the types in sync with the upstream module API as quickly as possible.

ctavan replied:
@rbuckton I have exported the V1Options and V4Options now. Do you think there's any value in exposing the sub-types that are used to compose these two types as well?

ctavan commented:
CI failure seems unrelated to this changeset.

ctavan commented:
Ready to merge
